realm/realm-library/src/main/cpp/jni_util/java_method.hpp

```diff
@@ -30,11 +31,28 @@ class JavaMethod {
         : m_method_id(nullptr)
     {
     }
-    JavaMethod(JNIEnv* env, jclass cls, const char* method_name, const char* signature, bool static_method = false);
-    JavaMethod(JNIEnv* env, jobject obj, const char* method_name, const char* signature);
-    JavaMethod(JNIEnv* env, const char* class_name, const char* method_name, const char* signature,
+    JavaMethod(JNIEnv* env, JavaClass const& cls, const char* method_name, const char* signature,
                bool static_method = false);
 
+    // From https://developer.android.com/training/articles/perf-jni.html
+    // The class references, field IDs, and method IDs are guaranteed valid until the class is unloaded. Classes are
+    // only unloaded if all classes associated with a ClassLoader can be garbage collected, which is rare but will not
+    // be impossible in Android. Note however that the jclass is a class reference and must be protected with a call
+    // to NewGlobalRef (see the next section).
+    //
```

Comment:
It is funny.  This is the first thing I thought of, when I saw all those static references to class objects.  Huh.